fn/goroutine_manager_test.go

```diff
@@ -0,0 +1,109 @@
+package fn
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+// TestGoroutineManager tests that GoroutineManager starts goroutines, until ctx
+// expires fails to start after it expires and waits for already started
+// goroutines in Stop method.
+func TestGoroutineManager(t *testing.T) {
+	t.Parallel()
+
+	m := NewGoroutineManager(context.Background())
+
+	taskChan := make(chan struct{})
+
+	require.NoError(t, m.Go(func(ctx context.Context) {
+		<-taskChan
+	}))
+
+	t1 := time.Now()
+
+	// Close taskChan in 1s, causing the goroutine to stop.
+	time.AfterFunc(time.Second, func() {
+		close(taskChan)
+	})
+
+	m.Stop()
+	stopDelay := time.Since(t1)
+
+	// Make sure Stop was waiting for the goroutine to stop.
+	require.Greater(t, stopDelay, time.Second)
+
+	// Make sure new goroutines do not start after Stop.
+	require.ErrorIs(t, m.Go(func(ctx context.Context) {}), ErrStopping)
+}
+
+// TestGoroutineManagerContextExpires tests the effect of context expiry.
+func TestGoroutineManagerContextExpires(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	m := NewGoroutineManager(ctx)
+
+	require.NoError(t, m.Go(func(ctx context.Context) {
+		<-ctx.Done()
+	}))
+
+	// The context of the manager should not expire, so the following call
+	// must block.
+	select {
+	case <-m.Context().Done():
+		t.Errorf("context must not expire at this point")
+	default:
+	}
+
+	cancel()
+
+	// The context of the manager should expire, so the following call
+	// must not block.
+	select {
+	case <-m.Context().Done():
```

Comment:
same here.

Reply:
`m.Context()` is not the same as `ctx`. `m.Context()` can expire earlier if `Stop()` is called. The intention of this code is to test `m.Context()`.

Also I added to TestGoroutineManager to make sure `m.Context()` expires after Stop is called.

Reply:
`m.Context()` should not exist on this API. Therefore it requires no tests.

Reply:
I replaced `Context()` method with `Done()` method returning `<-chan struct{}`.

Updated the test. Now it is testing `Done()` method.